Assignment 4/assignment-4-submission-folder/battle.cpp

```diff
@@ -0,0 +1,202 @@
+    switch (randIndex) {
+        case 0:
+            enemyType = Shrimp;
+            break;
+        case 1:
+            enemyType = Mosquito;
+            break;
+        case 2:
+            enemyType = Mimic;
+            break;
+    }
+    Enemy* enemy = get_enemy_of_type(enemyType);
+    cout << "\nYou encounter a " << enemy->name << "\n" << endl;
+
+    // Battle loop
+    while (true) {
+        // Players turn
+        for (int i = 0; i < party->num_players; i++) {
+            if (party->players[i].health <= 0) {
+                continue;
+            }
+            battle_turn(i, party, enemy);
+
+        // Check if enemy is defeated
+            if (enemy->health <= 0) {
+                cout << "\nEnemy Defeated!\n" << endl;
+                enemyDefeated = true;
+                break;
+            }
+        }
+
+        // Enemy's turn
+        int targetIndex;
+        do {
+            targetIndex = rand() % party->num_players;
+        } while (party->players[targetIndex].health <= 0);
+        // Damage
+        party->players[targetIndex].health -= enemy->damage;
+        cout << "The " << enemy->name << " attacks player " << targetIndex << " and deals " << enemy->damage << " damage." << endl;
+        // Check if player is alive
+        if (party->players[targetIndex].health <= 0) {
+            cout << "Player " << targetIndex << " has been defeated!" << endl;
+        }
+
+        //Check if all players are defeated
+        bool allPlayersDefeated = true;
+        for (int i = 0; i < party->num_players; i++) {
+            if (party->players[i].health > 0) {
+                allPlayersDefeated = false;
+                break;
+            }
+        }
+        if (allPlayersDefeated) {
+            cout << "All players have been defeated!" << endl;
+            break;
+        }
+    }
+    delete_enemy(enemy);
+}
+
+// Castle Battle
+void castle_battle(Party* party) {
+    // Check if enemy is defeated
+    if(!enemyDefeated) {
+        cout << "You must defeat an enemy in the mineshaft before you can enter the castle!" << endl;
+        return;
+    }
+
+    srand(time(nullptr));
+    Enemy* dragon = get_enemy_of_type(Dragon);
+    cout << "You enter the castle!\n" << endl;
+    cout << "The doors slam behind you, and you stand face to face witht he mighty dragon!\n" << endl;
+
+    // Battle loop
+    while (true) {
+        // Players turn
+        for (int i = 0; i < party->num_players; i++) {
+            if (party->players[i].health <= 0) {
+                continue;
+            }
+            battle_turn(i, party, enemy);
+            
+        // Check if enemy is defeated
+            if (enemy->health <= 0) {
+                cout << "\nEnemy Defeated!\n" << endl;
+                break;
+            }
+        }
+
+        // Dragon's turn
+        for (int i = 0; i < party->num_players; i++) {
+            if (party->players[i].health > 0) {
+                party->players[i].health -= Dragon->damage;
```

Comment:
dragon can inflict damage on deceased players; do with this information what you will (doesn't really matter than much, but you're free to change it if you want)